<html>
  <head>
    <title>Source Information for ens_35_segdup_sanger</title>
    <link rel="stylesheet" type="text/css" href="/stylesheets/das.css" />
  </head>
  <body>
    <h2>Source Information for ens_35_segdup_sanger</h2>
    <h3>Description</h3>
    <p>Segmental duplication dataset compiled by seeking homologies within the reference sequence using SSAHA2 and SEGMENT (unpublished) software.</p>

<p>The data were kindly provided by Nikolai Ivanov and Tony Cox at The 
Wellcome Trust Sanger Institute.</p>

<p>A region was considered a low copy repeat (LCR) if it had at least 90% 
identity to a different region of at least 3kb of the same chromosome.</p>

<p>Reference:<br>
Ning Z, Cox AJ, Mullikin JC. (2001) SSAHA: a fast search method for large 
DNA databases. Genome Res. 2001 Oct;11(10):1725-9. </p>

    <h3>Key</h3>
    <table>
	<tr>
		<th>Orientation</th>
		<td></td>
		<td></td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redvhigh.gif"><img src = "/gfx/style_lcr/redvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Red: Pair members in direct orientation</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenvhigh.gif"><img src = "/gfx/style_lcr/greenvhigh_l.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Green: Pair members in inverted orientation</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/blueboxvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Blue: Pair members on different chromosomes</td>
	</tr>
	<tr>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
	</tr>
	<tr>
		<th>% Identity</th>
		<td></td>
		<td></td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redvhigh.gif"><img src = "/gfx/style_lcr/greenvhigh.gif"><img src = "/gfx/style_lcr/blueboxvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Over 99% identity</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redhigh.gif"><img src = "/gfx/style_lcr/greenhigh.gif"><img src = "/gfx/style_lcr/blueboxhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>97 - 99% identity</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redmid.gif"><img src = "/gfx/style_lcr/greenmid.gif"><img src = "/gfx/style_lcr/blueboxmid.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>95 - 97% identity</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redlow.gif"><img src = "/gfx/style_lcr/greenlow.gif"><img src = "/gfx/style_lcr/blueboxlow.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Less than 95% identity</td>
	</tr>
	<tr>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
	</tr>
	<tr>
		<th>Spacing</th>
		<td></td>
		<td></td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Less than 100Kb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greennear.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>100Kb - 1Mb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenmed.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>1 - 5Mb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenfar.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>5 - 10Mb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenvfar.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>More than 10Mb apart</td>
	</tr>
      </table>


  </body>

</html>
