<html>
  <head>
    <title>Source Information for ens_35_segdup_washu</title>
    <link rel="stylesheet" type="text/css" href="/stylesheets/das.css" />
  </head>
  <body>
    <h2>Source Information for ens_35_segdup_washu</h2>
    <h3>Description</h3>
    <p>Segmental duplication dataset compiled by seeking homologies within the reference sequence.</p>
    <p>The data were kindly provided by Xinwei She and Evan Eichler at the University of Washington.</p>

<p>For a region to be included, at least 1 Kb of the total sequence (containing at least 500 bp of non-RepeatMasked sequence) had to align to another portion of the genome with a sequence identity of at least 90%.</p>

<p>Reference:<br>
Bailey JA, Gu Z, Clark RA, Reinert K, Samonte RV, Schwartz S, Adams MD, Myers EW, Li PW, Eichler EE (2002) Recent segmental duplications in the human genome. Science 297:1003-7.</p>
<p>Bailey JA, Yavor AM, Massa HF, Trask BJ, Eichler EE (2001) Segmental duplications: organization and impact within the current human genome project assembly. Genome Res 11:1005-17</p>
<p>She X, Jiang Z, Clark RA, Liu G, Cheng Z, Tuzun E, Church DM, Sutton G, Halpern AL, Eichler EE (2004) Shotgun sequence assembly and recent segmental duplications within the human genome. Nature 431:927-30</p>

    <h3>Key</h3>
    <table>
	<tr>
		<th>Orientation</th>
		<td></td>
		<td></td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redvhigh.gif"><img src = "/gfx/style_lcr/redvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Red: Pair members in direct orientation</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenvhigh.gif"><img src = "/gfx/style_lcr/greenvhigh_l.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Green: Pair members in inverted orientation</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/blueboxvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Blue: Pair members on different chromosomes</td>
	</tr>
	<tr>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
	</tr>
	<tr>
		<th>% Identity</th>
		<td></td>
		<td></td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redvhigh.gif"><img src = "/gfx/style_lcr/greenvhigh.gif"><img src = "/gfx/style_lcr/blueboxvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Over 99% identity</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redhigh.gif"><img src = "/gfx/style_lcr/greenhigh.gif"><img src = "/gfx/style_lcr/blueboxhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>97 - 99% identity</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redmid.gif"><img src = "/gfx/style_lcr/greenmid.gif"><img src = "/gfx/style_lcr/blueboxmid.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>95 - 97% identity</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/redlow.gif"><img src = "/gfx/style_lcr/greenlow.gif"><img src = "/gfx/style_lcr/blueboxlow.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Less than 95% identity</td>
	</tr>
	<tr>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
	</tr>
	<tr>
		<th>Spacing</th>
		<td></td>
		<td></td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenvhigh.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>Less than 100Kb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greennear.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>100Kb - 1Mb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenmed.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>1 - 5Mb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenfar.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>5 - 10Mb apart</td>
	</tr>
	<tr>
		<td><img src = "/gfx/style_lcr/greenvfar.gif"></td>
		<td>&nbsp;&nbsp;</td>
		<td>More than 10Mb apart</td>
	</tr>
      </table>


  </body>

</html>
